subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientForwardingProjectConfigurationOperationListener.java

```diff
@@ -39,18 +67,77 @@
+            } else if (buildOperation.getDetails() instanceof ApplyScriptPluginBuildOperationType.Details) {
+                ApplyScriptPluginBuildOperationType.Details details = (ApplyScriptPluginBuildOperationType.Details) buildOperation.getDetails();
+                incrementDuration(buildOperation, details.getTargetType(), details.getApplicationId(), finishEvent, () -> toScriptPluginIdentifier(details));
+            }
+            results.remove(buildOperation.getId());
+        }
+    }
+
     @Override
     protected InternalOperationStartedProgressEvent toStartedEvent(BuildOperationDescriptor buildOperation, OperationStartEvent startEvent, ConfigureProjectBuildOperationType.Details details) {
+        results.put(buildOperation.getId(), new ProjectConfigurationResult());
         return new DefaultOperationStartedProgressEvent(startEvent.getStartTime(), toProjectConfigurationDescriptor(buildOperation, details));
     }
 
     @Override
     protected InternalOperationFinishedProgressEvent toFinishedEvent(BuildOperationDescriptor buildOperation, OperationFinishEvent finishEvent, ConfigureProjectBuildOperationType.Details details) {
-        return new DefaultOperationFinishedProgressEvent(finishEvent.getEndTime(), toProjectConfigurationDescriptor(buildOperation, details), toOperationResult(finishEvent));
+        AbstractProjectConfigurationResult result = toProjectConfigurationOperationResult(finishEvent, results.remove(buildOperation.getId()));
+        return new DefaultOperationFinishedProgressEvent(finishEvent.getEndTime(), toProjectConfigurationDescriptor(buildOperation, details), result);
+    }
+
+    private InternalBinaryPluginIdentifier toBinaryPluginIdentifier(ApplyPluginBuildOperationType.Details details) {
+        String className = details.getPluginClass().getName();
+        String pluginId = details.getPluginId();
+        String displayName = MoreObjects.firstNonNull(pluginId, className);
+        return new DefaultBinaryPluginIdentifier(displayName, className, pluginId);
+    }
+
+    private InternalScriptPluginIdentifier toScriptPluginIdentifier(ApplyScriptPluginBuildOperationType.Details details) {
+        String fileString = details.getFile();
+        if (fileString != null) {
+            File file = new File(fileString);
+            return new DefaultScriptPluginIdentifier(file.getName(), file.toURI());
+        }
+        String uriString = details.getUri();
+        if (uriString != null) {
+            URI uri = URI.create(uriString);
+            return new DefaultScriptPluginIdentifier(FilenameUtils.getName(uri.getPath()), uri);
+        }
+        return null;
+    }
+
+    private void incrementDuration(BuildOperationDescriptor buildOperation, String targetType, long applicationId, OperationFinishEvent finishEvent, Supplier<? extends InternalPluginIdentifier> pluginSupplier) {
+        if (PROJECT_TARGET_TYPE.equals(targetType)) {
+            InternalPluginIdentifier plugin = pluginSupplier.get();
```

Comment:
I haven't benchmarked it but we do some `String` → `File` → `URI` conversions and `URI` parsing when dealing with script plugins.

Reply:
Given that project plugins are the majority anyway, I don't think this will make much of a difference.

Reply:
You're right, probably not. Are you concerned about potential overhead or the additional complexity?

Reply:
Just the complexity - It simply tripped me up for a second when I read the code. Nothing major, but I always ask in case I miss some important reason.